### DISCUSSION TITLE
How to choose the right URDF file?

### DISCUSSION THREAD

**Discussion by cedricgoubard** [robotology/community#546] [Q&A] [answered]
Hello, 

Is there any documentation explaining the various `urdf` versions that can be found in [the `iCub` models repository](https://github.com/robotology/icub-models)? How can I know which one fits our robot (iCubLondon01)? If the versioning is the same as the robot versions, is there anywhere I could find a `urdf` for an `iCub` 1.3?

Bonus question: is there an available model that includes `iKart`? I did not see any in the repository. Otherwise, I found a few `.sdf` files for `iKart` which I should be able to convert to `urdf`.

Thank you for your help!

**Comment by traversaro** [comment, marked as answer]
> Is there any documentation explaining the various urdf versions that can be found in the iCub models repository? How can I know which one fits our robot (iCubLondon01)? 

There is some documentation on the available models in https://github.com/robotology/icub-models-generator#generated-models, probably it could make sense to move it to `icub-models`, thanks for the observation. However, it is possible that none matches completly the iCubLondon01, so opening a discussion like you did is indeed a good idea, probably we should document this as well.

> If the versioning is the same as the robot versions, is there anywhere I could find a urdf for an iCub 1.3?

Are you sure it is a 1.3 ? Based on https://icub-tech-iit.github.io/documentation/icub_versions/ and on the data available in https://robot-bazaar.iit.it/robots/collection/robots/iCubLondon01, where the robot is marked as a `v1.x.1` without a precise indication on the minor version, I guess you should have a v2 neck (see https://icub-tech-iit.github.io/documentation/icub_tendons/neck/ for a quick difference between v1 and v2 neck). If indeed you have a neck v2 but you don't have skin on upper arm and chest but you have it on forearm and hands, indeed it is a v1.3.1 . If you can confirm this, it would be great so we can update the info on robot-bazaar.

Having said that, if indeed your robot has a v2 neck with v1 arms, an appropriate kinematic model for it may be https://github.com/robotology/icub-models-generator/tree/master/dh/data/legacy_models/headV2-armsV1-legsV1-feetV1, we can probably add a line in https://github.com/robotology/icub-models-generator/blob/v1.20.0/dh/CMakeLists.txt to generate it. However, that model does not include eyes and hands, so if you need them you may need to add them manually as they were added in https://github.com/robotology/icub-models/pull/42 . 


> Bonus question: is there an available model that includes `iKart`? I did not see any in the repository. Otherwise, I found a few `.sdf` files for `iKart` which I should be able to convert to `urdf`.

To be honest, I do not know anything regarding iKart URDF or SDF models. Perhaps @randaz81 may know more about this?

**Reply by cedricgoubard** [reply]
> Are you sure it is a 1.3 ? 

Not really, I just saw `1.3` mentioned a few times in discussions related to out `iCub` (*e.g.* [here](https://github.com/icub-tech-iit/documentation/pull/171#issuecomment-913445656)), but I'll trust your judgement.

>  If indeed you have a neck v2 but you don't have skin on upper arm and chest but you have it on forearm and hands, indeed it is a v1.3.1 

We have a v2 neck, but we do have skin on upper arms and chest: 
![PXL_20210908_090515541 MP](https://user-images.githubusercontent.com/40461909/132480624-fc89dcc8-a51a-4822-85a6-2c91ef07ca5c.jpg)
![PXL_20210908_084618517 MP](https://user-images.githubusercontent.com/40461909/132480640-e57023fd-447a-4bad-9483-29c179928d0d.jpg)

> To be honest, I do not know anything regarding iKart URDF or SDF models. 

No problem, thank you anyway. I'll let you know if I manage to merge our `iKart` files with the `iCub` file we'll end up choosing.

**Reply by traversaro** [reply]
Thanks for providing the photos! For sure the robot is at least a v1.4.1, but to be honest I want to double check internally if the arm are v1 or v2 (as the kinematics between the two is slightly different).

**Reply by cedricgoubard** [reply]
Hi @traversaro, 
Any news on the robot version? Could it help if I compared the arm with the schematics on the online documentation?

**Reply by traversaro** [reply]
Hi @cedricgoubard ! I checked with a few experts (@maggia80, @Fabrizio69, @davidetome, @pattacini) and they seem all to agree that indeed arms are indeed v1. So https://github.com/robotology/icub-models-generator/tree/master/dh/data/legacy_models/headV2-armsV1-legsV1-feetV1 may be the best model. Note that however it does not comes with eyes and fingers.

**Reply by cedricgoubard** [reply]
Hi @traversaro! Thank you for the file, I managed to include it with `iKart`:
![icub_urdf](https://user-images.githubusercontent.com/40461909/133239935-8aab8139-8250-4024-ab55-ffa28a682ea3.png)
